src/axom/primal/geometry/BezierPatch.hpp

```diff
@@ -0,0 +1,875 @@
+
+  /// Returns a copy of the Bezier patch's weights
+  axom::Array<T, 2> getWeights() const { return m_weights; }
+
+  /*!
+   * \brief Reverses the order of one direction of the Bezier patch's control points and weights
+   *
+   * \param [in] axis orientation of patch. 0 to reverse in u, 1 for reverse in v
+   */
+  void reverseOrientation(int axis)
+  {
+    const int ord_u = getOrder_u();
+    const int mid_u = (ord_u + 1) / 2;
+
+    const int ord_v = getOrder_v();
+    const int mid_v = (ord_v + 1) / 2;
+
+    if(axis == 0)
+    {
+      for(int q = 0; q <= ord_v; ++q)
+      {
+        for(int i = 0; i < mid_u; ++i)
+          axom::utilities::swap(m_controlPoints(i, q),
+                                m_controlPoints(ord_u - i, q));
+
+        if(isRational())
+          for(int i = 0; i < mid_u; ++i)
+            axom::utilities::swap(m_weights(i, q), m_weights(ord_u - i, q));
+      }
+    }
+    else
+    {
+      for(int p = 0; p <= ord_u; ++p)
+      {
+        for(int i = 0; i < mid_v; ++i)
+          axom::utilities::swap(m_controlPoints(p, i),
+                                m_controlPoints(p, ord_v - i));
+
+        if(isRational())
+          for(int i = 0; i < mid_v; ++i)
+            axom::utilities::swap(m_weights(p, i), m_weights(p, ord_v - i));
+      }
+    }
+  }
+
+  /// Swap the axes such that s(u, v) becomes s(v, u)
+  void swapAxes()
+  {
+    const int ord_u = getOrder_u();
+    const int ord_v = getOrder_v();
+
+    CoordsMat new_controlPoints(ord_v + 1, ord_u + 1);
+
+    for(int p = 0; p <= ord_u; ++p)
+      for(int q = 0; q <= ord_v; ++q)
+        new_controlPoints(q, p) = m_controlPoints(p, q);
+
+    m_controlPoints = new_controlPoints;
+
+    if(isRational())
+    {
+      axom::Array<T, 2> new_weights(ord_v + 1, ord_u + 1);
+      for(int p = 0; p <= ord_u; ++p)
+        for(int q = 0; q <= ord_v; ++q) new_weights(q, p) = m_weights(p, q);
+
+      m_weights = new_weights;
+    }
+  }
+
+  /// Returns an axis-aligned bounding box containing the Bezier patch
+  BoundingBoxType boundingBox() const
+  {
+    return BoundingBoxType(m_controlPoints.data(),
+                           static_cast<int>(m_controlPoints.size()));
+  }
+
+  /// Returns an oriented bounding box containing the Bezier patch
+  OrientedBoundingBoxType orientedBoundingBox() const
+  {
+    return OrientedBoundingBoxType(m_controlPoints.data(),
+                                   static_cast<int>(m_controlPoints.size()));
+  }
+
+  /*!
+   * \brief Evaluates a slice Bezier patch for a fixed parameter value of \a u or \a v
+   *
+   * \param [in] u parameter value at which to evaluate the first axis
+   * \param [in] v parameter value at which to evaluate the second axis
+   * \param [in] axis orientation of curve. 0 for fixed u, 1 for fixed v
+   * \return p the value of the Bezier patch at (u, v)
+   *
+   * \note We typically evaluate the patch at \a u or \a v between 0 and 1
+   */
+  BezierCurveType isocurve(T uv, int axis = 0) const
```

Comment:
Would this implementation and API be simpler w/ two seperate functions, e.g.: `isocurve_u()` and `isocurve_v()`?

Reply:
Given those two functions, we might even want to have a third w/ the current signature
```cpp
BezierCurveType isocurve(T uv, int axis = 0) const // returns either `isocurve_u()` or `isocurve_v()`
```

Reply:
If we're going to add an `isocurve_u` and `isocurve_v`, then I think we should do the same with `dt`,`split`, and `reverseOrientation`. Totally doable, but it would be odd to distinguish for one and not the rest.

Reply:
However we do it though, this shouldn't be a default parameter. I don't like the idea of giving one axis any privilege over the other.